experimental/Origami/src/generate.jl

```diff
@@ -0,0 +1,412 @@
+# TODO: correct typing
+
+# the bijection between the cycles of the bot/top permutation
+# is given implicitly by using lists and considering the order
+# of the cycles
+struct CylinderDiagram
+    bot::Vector{Vector{Int}}
+    top::Vector{Vector{Int}}
+    cycles_count::Int
+    separatrix_count::Int
+    function CylinderDiagram(bot::Vector{Vector{Int}}, top::Vector{Vector{Int}})
+        max::Int = 1;
+        for cycle in bot
+            for i in cycle
+                if(i > max)
+                    max = i
+                end
+            end
+        end
+        new(bot, top, length(bot), max + 1)
+    end
+end
+
+function cylinders(cylinder_diagram::CylinderDiagram)::Vector{Vector{Vector{Int}}}
+    result = Vector{Vector{Vector{Int}}}()
+    for i in 1:cylinder_diagram.cycles_count
+        push!(result, [cylinder_diagram.bot[i], cylinder_diagram.top[i]])
+    end
+    return result
+end
+
+# computes the system of equations whose solutions are the realizable cylinder
+# coordinates
+function system_of_equations(cylinder_diagram::CylinderDiagram)
+    M = zero_matrix(ZZ, cylinder_diagram.cycles_count, cylinder_diagram.separatrix_count)
+    for (i, (bot, top)) in enumerate(cylinders(cylinder_diagram))
+        for t in top
+            M[i, t + 1] = 1
+        end
+        for b in bot
+            M[i, b + 1] -= 1
+        end
+    end
+    return M
+end
+
+# positive integer linear combinations of the computed rays
+# give possible lengths for the separatrix of the cylinder diagram
+# i.e. the resulting surface exists
+function compute_rays(equations, separatrix_count::Int)::Vector{Vector{Int}}
+    if(is_zero(equations))
+        n = separatrix_count
+        return [Int[i == j for j in 1:n] for i in 1:n]
+    end
+    neg_identity = -identity_matrix(ZZ, separatrix_count)
+    cone = cone_from_inequalities(neg_identity, equations)
+    ray_list = rays(cone)
+    return map(inner -> map(Int, inner), ray_list)
+end
+
+# generate all possible linear combinations
+# m: number of parameters in linear combination, parameters have integer values from 1 to n
+function all_combinations(n::Int, m::Int)::Vector{Vector{Int}}
+    combinations = Vector{Vector{Int}}([[]])
+
+    for _ in 1:m
+        new_combinations = Vector{Vector{Int}}()
+        for current in combinations
+            for i in 1:n
+                push!(new_combinations, [current..., i])
+            end
+        end
+        combinations = new_combinations
+    end
+    return combinations
+end
```

Comment:
This function should be replaced by a call to `tuples(n,m)` -- well, once it is implemented, see https://github.com/oscar-system/Oscar.jl/issues/5360